src/core/lib/channel/deadline_filter.c

```diff
@@ -80,11 +78,15 @@ static void start_timer_if_needed(grpc_exec_ctx* exec_ctx,
       (grpc_deadline_timer_state)gpr_atm_acq_load(&deadline_state->timer_state);
   switch (cur_state) {
     case GRPC_DEADLINE_STATE_PENDING:
+      // Note: We do not start the timer if there is already a timer
       return;
     case GRPC_DEADLINE_STATE_FINISHED:
       if (gpr_atm_rel_cas(&deadline_state->timer_state,
                           GRPC_DEADLINE_STATE_FINISHED,
                           GRPC_DEADLINE_STATE_PENDING)) {
+        // If we've already created and destroyed a timer, we always create a
```

Comment:
The closure contains a next pointer, so we need to be guaranteed that the callback has executed (not just triggered) before reusing it (I think this bug existed previously also).